File: 1463.cpp
#pragma warning(disable:4996)
#include <iostream>
using namespace std;

int min(int a, int b) {
	return (a < b) ? a : b;
}

int main() {

	int N;
	cin >> N;

	int* cnts = new int[N + 1];
	cnts[N] = 0;
	for (int i = 0; i < N; i++)
		cnts[i] = 999999;

	for (int i = N; i > 0; i--) {
		int tmp;
		if (i % 3 == 0) {
			tmp = i / 3;
			cnts[tmp] = min(cnts[i] + 1, cnts[tmp]);
		}
		if (i % 2 == 0) {
			tmp = i / 2;
			cnts[tmp] = min(cnts[i] + 1, cnts[tmp]);
		}
		tmp = i - 1;
		cnts[tmp] = min(cnts[i] + 1, cnts[tmp]);
	}

	cout << cnts[1];

	return 0;
}

File: 1463.py
N = int(input())

dp = [0, 0, 1, 1]
for i in range(4, N+1):
    tmp = float("inf")
    if i%3 == 0:
        tmp = min(dp[i//3]+1, dp[i-1]+1)

    if i%2 == 0:
        tmp = min(tmp, (min(dp[i//2]+1, dp[i-1]+1)))
    
    dp.append(min(tmp, dp[i-1]+1))

print(dp[N])
